File: problems/leetcode/0400/416/416.cpp
// 416. Partition Equal Subset Sum
// https://leetcode.com/problems/partition-equal-subset-sum/description/

using namespace std;

#include <map>
#include <numeric>
#include <vector>

class Solution {
public:
  bool dfs(vector<int> &nums, int i, int current, int target,
           map<pair<int, int>, bool> &memo) {
    if (memo.find({i, current}) != memo.end()) {
      return memo[{i, current}];
    }

    if (current > target) {
      return false;
    }

    if (i == nums.size()) {
      return current == target;
    }

    bool available = dfs(nums, i + 1, current + nums[i], target, memo) ||
                     dfs(nums, i + 1, current, target, memo);
    memo[{i, current}] = available;
    return available;
  }

  bool canPartition(vector<int> &nums) {
    int total = accumulate(nums.begin(), nums.end(), 0);
    if (total % 2 != 0) {
      return false;
    }
    int target = total / 2;

    // (index, sum)
    map<pair<int, int>, bool> memo;
    return dfs(nums, 0, 0, target, memo);
  }
};

File: problems/leetcode/0400/416/416.py
# 416. Partition Equal Subset Sum
# https://leetcode.com/problems/partition-equal-subset-sum/description/

# 1. Naive Solution
# 일단 합이 홀수이면 불가능
# Backtracking

# 2. Solution
# Knapsack Approach
# dp[i][j]: i번째 index에서 합이 j가 될 수 있는지
# dp[i][j] = dp[i - 1][j] || dp[i - 1][j - nums[i]]
# 합이 전체의 절반이 되면 됨.

from typing import *

class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        n = len(nums)

        total = sum(nums)
        if total % 2 == 1:
            return False

        memo = [False for _ in range((total // 2) + 1)]
        memo[0] = True

        for i in range(n):
            newMemo = [prev for prev in memo]

            for j in range((total // 2) + 1):
                if j - nums[i] >= 0:
                    newMemo[j] |= memo[j - nums[i]]

            memo = newMemo

        return memo[total // 2]
